megatron/model/gpt_model.py

```diff
@@ -338,36 +339,88 @@ def _logits_helper(embedding, lm_output):
                          activation_checkpoint_interval=interval,
                          partition_method='type:transformer')
 
+    @staticmethod
+    def _get_vocab_param_patterns():
+        args = get_args()
+        if args.untie_embeddings_and_output_weights:
+            patterns = [
+                r"\d+.word_embeddings.weight",
+                r"\d+.lm_head.weight"
+            ]
+        else:
+            patterns = [
+                r"tied_modules.embed.word_embeddings.weight"
+            ]
+        return patterns
+
+    def _get_pp_replicated_param_patterns(self):
+        args = get_args()
+        if args.untie_embeddings_and_output_weights:
+            return []
+        patterns = self._get_vocab_param_patterns()
+        if args.add_position_embedding:
+            patterns.append(r"tied_modules.embed.position_embeddings.weight")
+        return patterns
+
+    @staticmethod
+    def _get_tp_replicated_param_patterns():
+        args = get_args()
+        patterns = [
+            r"\d+.input_layernorm.weight",
+            r"\d+.post_attention_layernorm.weight",
+            r"\d+.weight",
+        ]
+        if args.add_position_embedding:
+            patterns.append(r"tied_modules.embed.position_embeddings.weight")
+        if args.add_bias_linear:
+            patterns.extend([
+                r"\d+.self_attention.dense.bias",
+                r"\d+.mlp.dense_4h_to_h.bias",
+            ])
+        if args.normalization == 'layernorm':
+            patterns.extend([
+                r"\d+.input_layernorm.bias",
+                r"\d+.post_attention_layernorm.bias",
+                r"\d+.bias",
+            ])
+        return patterns
+
+    @staticmethod
+    def _get_row_parallel_param_patterns():
+        return [
+            r"\d+.mlp.dense_4h_to_h.weight",
+            r"\d+.self_attention.dense.weight",
+        ]
+
+    @staticmethod
+    def _get_swiglu_col_parallel_param_patterns():
```

Comment:
Can you please confirm that `--swiglu` command line option is all that is needed to turn a gpt model into a llama model? I ask because I was expecting to see these changes in a `megatron/model/llama_model.py` rather than `megatron/model/gpt_model.py` 

I am not familiar with the llama support in this code base.

Reply:
LLAMA architecture was integrated into gpt_model, and I don't think there is `megatron/model/llama_model.py`.
In high level, LLAMA requires the following "architecture-related" flags:
```
--no-query-key-layer-scaling
--use-rotary-position-embeddings
--untie-embeddings-and-output-weights
--swiglu
--normalization rmsnorm
--disable-bias-linear 
```
Actually, I used the following existing script as a base for my testing: `examples_deepspeed/pretrain_llama2_distributed.sh`

During my testing, when trying to run llama scripts, I encountered few small issues that I had to fix.
They are not related to universal, so not included in this PR.